tests/TransactionBuilderTest.php

```diff
@@ -0,0 +1,220 @@
+        DB::commit();
+        $user = DB::collection('users')->where('name', $this->originData['name'])->first();
+        $this->assertEquals($this->originData['age'] + 1, $user['age']);
+    }
+
+    public function testDecrement()
+    {
+        /** rollback test */
+        DB::beginTransaction();
+        DB::collection('users')->where('name', $this->originData['name'])->decrement('age');
+        DB::rollBack();
+        $user = DB::collection('users')->where('name', $this->originData['name'])->first();
+        $this->assertEquals($this->originData['age'], $user['age']);
+
+        /** commit test */
+        DB::beginTransaction();
+        DB::collection('users')->where('name', $this->originData['name'])->decrement('age');
+        DB::commit();
+        $user = DB::collection('users')->where('name', $this->originData['name'])->first();
+        $this->assertEquals($this->originData['age'] - 1, $user['age']);
+    }
+
+    public function testQuery()
+    {
+        /** rollback test */
+        DB::beginTransaction();
+        $count = DB::collection('users')->count();
+        $this->assertEquals(1, $count);
+        DB::collection('users')->insert($this->insertData);
+        $count = DB::collection('users')->count();
+        $this->assertEquals(2, $count);
+        DB::rollBack();
+        $count = DB::collection('users')->count();
+        $this->assertEquals(1, $count);
+
+        /** commit test */
+        DB::beginTransaction();
+        $count = DB::collection('users')->count();
+        $this->assertEquals(1, $count);
+        DB::collection('users')->insert($this->insertData);
+        $count = DB::collection('users')->count();
+        $this->assertEquals(2, $count);
+        DB::commit();
+        $count = DB::collection('users')->count();
+        $this->assertEquals(2, $count);
+    }
+
+    public function testTransaction()
+    {
+        $count = DB::collection('users')->count();
+        $this->assertEquals(1, $count);
+
+        $new_age = $this->originData['age'] + 1;
+        DB::transaction(function () use ($new_age) {
+            DB::collection('users')->insert($this->insertData);
+            DB::collection('users')->where('name', $this->originData['name'])->update(['age' => $new_age]);
+        });
+        $count = DB::collection('users')->count();
+        $this->assertEquals(2, $count);
+
+        $checkInsert = DB::collection('users')->where('name', $this->insertData['name'])->first();
+        $this->assertNotNull($checkInsert);
+        $this->assertEquals($this->insertData['age'], $checkInsert['age']);
+
+        $checkUpdate = DB::collection('users')->where('name', $this->originData['name'])->first();
+        $this->assertEquals($new_age, $checkUpdate['age']);
+    }
+
+    /**
+     * Supports infinite-level nested transactions, but outside transaction rollbacks do not affect the commit of inside transactions
```

Comment:
I still don't understand how this works. Assuming this all uses the same Session object, I expect we'd get an error when WithTransaction attempts to call `Session::startTransaction()`.

Does this test actually pass?

Reply:
That is due to the wrong implementation of `DB::transaction`, which does not re-use an existing session object on the transaction. Once that is resolved, this test will be failing with the same `RuntimeException` expected in the `TransactionTest` when doing two subsequent calls to `DB::startTransaction()`.

Reply:
I think this test was never ran because it start with the capital letter "T". In any case, I will modify this test a little bit, so I will make all the needed corrections.